flask/src/main/java/net/corda/gradle/flask/FlaskJarTask.java

```diff
@@ -0,0 +1,235 @@
+import lombok.Setter;
+import lombok.SneakyThrows;
+import net.corda.flask.common.Flask;
+import net.corda.flask.common.ManifestEscape;
+import org.gradle.api.Action;
+import org.gradle.api.GradleException;
+import org.gradle.api.file.FileCollection;
+import org.gradle.api.file.FileCopyDetails;
+import org.gradle.api.internal.file.CopyActionProcessingStreamAction;
+import org.gradle.api.internal.file.copy.CopyAction;
+import org.gradle.api.internal.file.copy.CopyActionProcessingStream;
+import org.gradle.api.internal.file.copy.FileCopyDetailsInternal;
+import org.gradle.api.model.ObjectFactory;
+import org.gradle.api.plugins.BasePluginConvention;
+import org.gradle.api.provider.ListProperty;
+import org.gradle.api.provider.Property;
+import org.gradle.api.tasks.Input;
+import org.gradle.api.tasks.InputFiles;
+import org.gradle.api.tasks.WorkResult;
+import org.gradle.api.tasks.bundling.AbstractArchiveTask;
+
+import java.io.File;
+import java.io.FileInputStream;
+import java.io.FileOutputStream;
+import java.io.InputStream;
+import java.math.BigInteger;
+import java.security.DigestInputStream;
+import java.security.MessageDigest;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Objects;
+import java.util.Optional;
+import java.util.function.Function;
+import java.util.jar.Attributes;
+import java.util.jar.JarFile;
+import java.util.jar.Manifest;
+import java.util.stream.Collectors;
+import java.util.zip.ZipEntry;
+import java.util.zip.ZipOutputStream;
+
+public class FlaskJarTask extends AbstractArchiveTask {
+
+    @Setter
+    private Property<String> launcherClassName;
+
+    @Input
+    public Property<String> getLauncherClassName() {
+        return launcherClassName;
+    }
+
+    @Setter
+    private Property<String> mainClassName;
+
+    @Input
+    public Property<String> getMainClassName() {
+        return mainClassName;
+    }
+
+    @Setter
+    private ListProperty<String> jvmArgs;
+
+    @Input
+    public ListProperty<String> getJvmArgs() {
+        return jvmArgs;
+    }
+
+    public void includeLibraries(FileCollection fileCollection) {
+        into(Flask.Constants.LIBRARIES_FOLDER, (copySpec) -> copySpec.from(fileCollection));
+    }
+
+    private static class JavaAgent {
+        File jar;
+        String args;
+    }
+
+    private final List<JavaAgent> javaAgents;
+
+    @InputFiles
+    FileCollection getAgentJars() {
+        File[] jars = javaAgents.stream().map(it -> it.jar).toArray(File[]::new);
+        return getProject().files((Object[]) jars);
+    }
+
+    @Input
+    List<String> getAgentArgs() {
+        return javaAgents.stream().map(it -> it.args).collect(Collectors.toList());
+    }
+
+    public void javaAgent(Action<JavaAgent> action) {
+        JavaAgent agent = new JavaAgent();
+        action.execute(agent);
+        if(agent.jar == null) throw new GradleException("No jar file specified for Java agent");
+        javaAgents.add(agent);
+    }
+
+    public FlaskJarTask() {
```

Comment:
Gradle supports dependency injection, to a certain degree. It's better to use it rather than relying on `Project`:
```java
@Inject
public FlaskJarTask(ObjectFactory objects) {
   ...
}
```

Reply:
ack